<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Payam Emami">

<title>Multi-Omics Factor Analysis (MOFA)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="mofa-light_files/libs/clipboard/clipboard.min.js"></script>
<script src="mofa-light_files/libs/quarto-html/quarto.js"></script>
<script src="mofa-light_files/libs/quarto-html/popper.min.js"></script>
<script src="mofa-light_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="mofa-light_files/libs/quarto-html/anchor.min.js"></script>
<link href="mofa-light_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="mofa-light_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="mofa-light_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="mofa-light_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="mofa-light_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Multi-Omics Factor Analysis (MOFA)</h1>
            <p class="subtitle lead">Light Version</p>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Payam Emami </p>
            </div>
    </div>
      
    
      
    </div>
    
  
  </header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="3">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data</a></li>
  <li><a href="#data-integration-using-mofa2" id="toc-data-integration-using-mofa2" class="nav-link" data-scroll-target="#data-integration-using-mofa2">Data integration using MOFA2</a>
  <ul>
  <li><a href="#building-a-mofa-object" id="toc-building-a-mofa-object" class="nav-link" data-scroll-target="#building-a-mofa-object">Building a MOFA object</a></li>
  <li><a href="#defining-data-and-model-options" id="toc-defining-data-and-model-options" class="nav-link" data-scroll-target="#defining-data-and-model-options">Defining data and model options</a></li>
  <li><a href="#training-a-mofa-object" id="toc-training-a-mofa-object" class="nav-link" data-scroll-target="#training-a-mofa-object">Training a MOFA object</a></li>
  <li><a href="#variance-decomposition" id="toc-variance-decomposition" class="nav-link" data-scroll-target="#variance-decomposition">Variance decomposition</a></li>
  <li><a href="#downstream-analysis" id="toc-downstream-analysis" class="nav-link" data-scroll-target="#downstream-analysis">Downstream analysis</a></li>
  <li><a href="#adding-annotations" id="toc-adding-annotations" class="nav-link" data-scroll-target="#adding-annotations">Adding annotations</a>
  <ul class="collapse">
  <li><a href="#visualizing-factors" id="toc-visualizing-factors" class="nav-link" data-scroll-target="#visualizing-factors">Visualizing factors</a></li>
  <li><a href="#association-analysis" id="toc-association-analysis" class="nav-link" data-scroll-target="#association-analysis">Association analysis</a></li>
  <li><a href="#visualizing-weights" id="toc-visualizing-weights" class="nav-link" data-scroll-target="#visualizing-weights">Visualizing weights</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">




<p>Setting up environment</p>
<p>You will need to install a few packages to fully run this notebook. The main packages needed are <code>MOFA2</code> and <code>ggplot2</code>.</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-1_dbf1d7eb93b420a9a6d995d9dd47f887">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"MOFA2"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># list of packages to be installed</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>packages <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># check and install missing packages</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>new_packages <span class="ot">&lt;-</span> packages[<span class="sc">!</span>(packages <span class="sc">%in%</span> <span class="fu">installed.packages</span>()[,<span class="st">"Package"</span>])]</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">length</span>(new_packages)) <span class="fu">install.packages</span>(new_packages, <span class="at">dependencies =</span> <span class="cn">TRUE</span>, <span class="at">type =</span> <span class="st">"binary"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>In classical data integration, we would like to use information across different modalities (e.g.&nbsp;transcriptome, proteome and metabolome) to gain more comprehensive insights into the biological systems under study. This type of data can be used for an array of different purposes including but not limited to molecular classification, stratification of patients, outcome predictions and understanding of regulatory processes such as gene regulation and pathway analysis.</p>
<p>Here we are going to focus on unsupervised modeling and segmentation, which are promising because each type of omics data may contribute valuable information to the overall understanding of complex biological systems. By leveraging unsupervised modeling, we can uncover hidden patterns and relationships within the data without relying on predefined labels. This is especially beneficial when dealing with omics data, where the volume and complexity can be overwhelming. Furthermore, segmentation allows us to group similar data points, making it easier to identify and analyze specific subsets of the data. Given the heterogeneous nature of omics data, integrating multiple types can provide a more comprehensive view of the underlying biological processes.</p>
<p>In this lab we are going to learn how to use Multi-Omics Factor Analysis (MOFA) for multiple data views to uncover hidden but common pattern within the data.</p>
</section>
<section id="data" class="level1">
<h1>Data</h1>
<p>We will use TCGA data set from <code>mixOmics</code>, where features are in the columns and samples in the rows.</p>
<blockquote class="blockquote">
<p><em>This data set is a small subset of the full data set from The Cancer Genome Atlas. It contains the expression or abundance of three matching omics data sets: mRNA, miRNA and proteomics for 150 breast cancer samples (with three molecular subtypes of breast cancer: Basal, Her2, Luminal A) in the training set, and 70 samples in the test set. The test set is missing the proteomics data set.</em></p>
</blockquote>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-2_4c020f0ab935b1064e3057d0566ab02b">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download the dataset</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://github.com/mixOmicsTeam/mixOmics/raw/master/data/breast.TCGA.rda"</span>, <span class="at">destfile =</span> <span class="st">"TCGA.rda"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># load the data</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"TCGA.rda"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This data, when loaded, has already been split into a list with two elements: training and testing. The first element (training) contains four elements, again lists, containing miRNA, mRNA, proteomics and cancer molecular subtypes. The second element (testing) contains three lists holding miRNA, mRNA and molecular subtypes data (proteomics data are missing here).</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-3_675c7ddcdd80d90d33d41cf28682d01f">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># preview data</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(breast.TCGA)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 2
 $ data.train:List of 4
  ..$ mirna  : num [1:150, 1:184] 11.8 12.9 12.3 12 13.4 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:150] "A0FJ" "A13E" "A0G0" "A0SX" ...
  .. .. ..$ : chr [1:184] "hsa-let-7a-1" "hsa-let-7a-2" "hsa-let-7a-3" "hsa-let-7b" ...
  ..$ mrna   : num [1:150, 1:200] 4.36 1.98 1.73 4.36 2.45 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:150] "A0FJ" "A13E" "A0G0" "A0SX" ...
  .. .. ..$ : chr [1:200] "RTN2" "NDRG2" "CCDC113" "FAM63A" ...
  ..$ protein: num [1:150, 1:142] 0.0491 -0.08 -0.0328 -0.2053 0.0602 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:150] "A0FJ" "A13E" "A0G0" "A0SX" ...
  .. .. ..$ : chr [1:142] "14-3-3_epsilon" "4E-BP1" "4E-BP1_pS65" "4E-BP1_pT37" ...
  ..$ subtype: Factor w/ 3 levels "Basal","Her2",..: 1 1 1 1 1 1 1 1 1 1 ...
 $ data.test :List of 3
  ..$ mirna  : num [1:70, 1:184] 12.8 13.9 12.9 12.4 13.1 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:70] "A54N" "A2NL" "A6VY" "A3XT" ...
  .. .. ..$ : chr [1:184] "hsa-let-7a-1" "hsa-let-7a-2" "hsa-let-7a-3" "hsa-let-7b" ...
  ..$ mrna   : num [1:70, 1:200] 1.19 2.73 3.05 2.7 3.14 ...
  .. ..- attr(*, "dimnames")=List of 2
  .. .. ..$ : chr [1:70] "A54N" "A2NL" "A6VY" "A3XT" ...
  .. .. ..$ : chr [1:200] "RTN2" "NDRG2" "CCDC113" "FAM63A" ...
  ..$ subtype: Factor w/ 3 levels "Basal","Her2",..: 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
</div>
</div>
</section>
<section id="data-integration-using-mofa2" class="level1">
<h1>Data integration using MOFA2</h1>
<p>MOFA is a factor analysis model that provides a general framework for the integration of multi-omic data sets in an unsupervised fashion. Intuitively, MOFA can be viewed as a versatile and statistically rigorous generalization of principal component analysis to multi-omics data. Given several data matrices with measurements of multiple -omics data types on the same or on overlapping sets of samples, MOFA infers an interpretable low-dimensional representation in terms of a few latent factors. These learnt factors represent the driving sources of variation across data modalities, thus facilitating the identification of cellular states or disease subgroups.</p>
<p>Let’s use our omics data, mRNA, miRNA and proteome to create a low-dimensional representation based on the variation across the three modalities. To do that, we will need to build and train MOFA object. After that, in downstream analysis we can use the MOFA model to visualize and interpret the model output, e.g.&nbsp;to learn how much variance is explained by the individual omics for the new latent factors or find out which features relate most (have strongest features weights) to the latent factors.</p>
<section id="building-a-mofa-object" class="level2">
<h2 class="anchored" data-anchor-id="building-a-mofa-object">Building a MOFA object</h2>
<p>To create a MOFA object you need to specify three dimensions: samples, features and view(s). Views(s) are the different omics types in our case. MOFA object can be created based on many different data formats such as a list of matrices, a long data.frame, MultiAssayExperiment or even Suerat objects for single-cell genomics data. Here, we will use a list of matrices as our TCGA data is already in this format.</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-4_6de4e31aae80a70e7abe82b6781b2849">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load library</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MOFA2)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the subtype information from training data</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>data_mofa <span class="ot">&lt;-</span> breast.TCGA<span class="sc">$</span>data.train[<span class="sc">-</span><span class="dv">4</span>]</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose data because MOFA wants features in rows</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>data_mofa <span class="ot">&lt;-</span> <span class="fu">lapply</span>(data_mofa, t)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># create MOFA object</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>MOFAobject <span class="ot">&lt;-</span> <span class="fu">create_mofa</span>(data_mofa)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="do">## Creating MOFA object from a list of matrices (features as rows, sample as columns)...</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can have a look at the structure of the input data:</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-5_155b204b9a966e82972e50dcc3c75c8c">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_data_overview</span>(MOFAobject)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This shows us how many samples we have and how many features per data view is there. If there were missing values, these would be shown as gray lines.</p>
<div id="exr-na" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 1 (NA) </strong></span>Can you add a bit of missing values and check how the plot will change?</p>
</div>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-6_e6b9f33088ea359f74b0e463a73e075a">
<details>
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Let's randomly introduce NAs to 20% of samples in one omics e.g. protein data</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This is an example code. You can probably find an easier way to solve this :) </span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># make a copy of MOFA data and protein data</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>data_mofa_with_na <span class="ot">&lt;-</span> data_mofa</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>data_protein <span class="ot">&lt;-</span> data_mofa<span class="sc">$</span>protein</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate number of data points to replace</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">ncol</span>(data_protein) <span class="co"># no. of samples</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>n_to_replace <span class="ot">&lt;-</span> <span class="dv">20</span><span class="sc">/</span><span class="dv">100</span> <span class="sc">*</span> n <span class="co"># number to replace, 20%</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a><span class="co"># sample index and replace with NA </span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>data_protein[, <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>n, n_to_replace)] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="co"># check that we have NAs, we should have n_to_replace amount</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="co"># sum(is.na(data_protein))</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a><span class="co"># replace protein data under the MOFA</span></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>data_mofa_with_na<span class="sc">$</span>protein <span class="ot">&lt;-</span> data_protein</span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a><span class="co"># create MOFA object</span></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a>MOFAobject_with_na <span class="ot">&lt;-</span> <span class="fu">create_mofa</span>(data_mofa_with_na)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Creating MOFA object from a list of matrices (features as rows, sample as columns)...</code></pre>
</div>
<details>
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_data_overview</span>(MOFAobject_with_na)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="defining-data-and-model-options" class="level2">
<h2 class="anchored" data-anchor-id="defining-data-and-model-options">Defining data and model options</h2>
<p>Before we start modeling, we can specify some data and model options.</p>
<p>For <strong>data options</strong>, we have:</p>
<ul>
<li>scale_groups: if groups have different ranges/variances, it is good practice to scale each group to unit variance. Default is FALSE</li>
<li>scale_views: if views have different ranges/variances, it is good practice to scale each view to unit variance. Default is FALSE</li>
</ul>
<p>and we can confirm that we are using default options by:</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-7_4ca0eccb5d648aea5a21aa7440ae6179">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>data_opts <span class="ot">&lt;-</span> <span class="fu">get_default_data_options</span>(MOFAobject)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(data_opts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$scale_views
[1] FALSE

$scale_groups
[1] FALSE

$center_groups
[1] TRUE

$use_float32
[1] TRUE

$views
[1] "mirna"   "mrna"    "protein"

$groups
[1] "group1"</code></pre>
</div>
</div>
<p>For <strong>model options</strong>, we have:</p>
<ul>
<li>num_factors: number of factors</li>
<li>likelihoods: likelihood per view (options are “gaussian”, “poisson”, “bernoulli”). Default is “gaussian”.</li>
<li>spikeslab_factors: use spike-slab sparsity prior in the factors? Default is FALSE.</li>
<li>spikeslab_weights: use spike-slab sparsity prior in the weights? Default is TRUE.</li>
<li>ard_factors: use ARD prior in the factors? Default is TRUE if using multiple groups.</li>
<li>ard_weights: use ARD prior in the weights? Default is TRUE if using multiple views.</li>
</ul>
<p>We can control the number of factors and we should adjust the likelihoods to match our data. Unless we want to learn more about the underlying mathematical models, we keep other parameters, such as spikeslab and ARD priors set to default.</p>
<p>Let’s check our omics data distributions to make sure we use correct likelihood values.</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-8_8ae1bf989b7fa601b4d6f71a9d9e536b">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(data_mofa<span class="sc">$</span>mrna)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(data_mofa<span class="sc">$</span>protein)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(data_mofa<span class="sc">$</span>mirna)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>All of our data is seems to be normally distributed so we use normal distribution. In practice MOFA allows us to select ‘gaussian’ for continuous data (e.g proteomics), ‘bernoulli’ for binary data (e.g.&nbsp;methylation) and ‘poisson’ for count data (e.g.&nbsp;RNA-Seq).</p>
<p>We can now set the model parameters. We can preview the default parameters already set:</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-9_950378f5f64ef88c03fd6266a6e4a509">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>model_opts <span class="ot">&lt;-</span> <span class="fu">get_default_model_options</span>(MOFAobject)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(model_opts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$likelihoods
     mirna       mrna    protein 
"gaussian" "gaussian" "gaussian" 

$num_factors
[1] 15

$spikeslab_factors
[1] FALSE

$spikeslab_weights
[1] FALSE

$ard_factors
[1] FALSE

$ard_weights
[1] TRUE</code></pre>
</div>
</div>
<p>where we see that MOFA selected default (gaussian) likelihood for all our data and includes 15 factors (latent variables).</p>
<p>To change model parameters, e.g.&nbsp;reduce number of factors from default 15 to 10 to make the computations run faster we type:</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-10_3400789a95b7eb617ed28ab7e156e1ca">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>model_opts<span class="sc">$</span>num_factors <span class="ot">&lt;-</span> <span class="dv">10</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="training-a-mofa-object" class="level2">
<h2 class="anchored" data-anchor-id="training-a-mofa-object">Training a MOFA object</h2>
<p>Our MOFA object is now set and we can start the training. Similar to model options, there are parameters that define training options. Briefly, these are:</p>
<ul>
<li>maxiter: number of iterations</li>
<li>convergence_mode</li>
<li>gpu_mode</li>
<li>verbose mode</li>
</ul>
<p>We can again see the default values:</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-11_a73364049f35575ee6f0ed1675fdc146">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>train_opts <span class="ot">&lt;-</span> <span class="fu">get_default_training_options</span>(MOFAobject)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(train_opts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$maxiter
[1] 1000

$convergence_mode
[1] "fast"

$drop_factor_threshold
[1] -1

$verbose
[1] FALSE

$startELBO
[1] 1

$freqELBO
[1] 5</code></pre>
</div>
</div>
<p>and notice that for instance that the default number of iterations is set to 1000 and the convergence mode is set to “fast”. Similar to model options, these parameters refer to the underlying method. “Fast” convergence mode tends to be good for exploration, but it may be worth considering changing it to “medium” or “slow” for the final model. GPU mode refers to running MOFA on GPU, something that needs cupy installed and a functional GPU.</p>
<p>To train a MOFA object:</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-12_5c838f339711455c6a6eccc8be15a914">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>MOFAobject <span class="ot">&lt;-</span> <span class="fu">prepare_mofa</span>(MOFAobject,</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">model_options =</span> model_opts <span class="co"># input model options</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Checking data options...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>No data options specified, using default...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>No training options specified, using default...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Checking model options...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>MOFAobject <span class="ot">&lt;-</span> <span class="fu">invisible</span>(<span class="fu">run_mofa</span>(MOFAobject, <span class="at">use_basilisk =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in run_mofa(MOFAobject, use_basilisk = TRUE): No output filename provided. Using /var/folders/hw/jx67_4vj6ljfd13xsg7xzvt83k7mrx/T//Rtmplep0xV/mofa_20231105-162058.hdf5 to store the trained model.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Connecting to the mofapy2 package using basilisk. 
    Set 'use_basilisk' to FALSE if you prefer to manually set the python binary using 'reticulate'.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
        #########################################################
        ###           __  __  ____  ______                    ### 
        ###          |  \/  |/ __ \|  ____/\    _             ### 
        ###          | \  / | |  | | |__ /  \ _| |_           ### 
        ###          | |\/| | |  | |  __/ /\ \_   _|          ###
        ###          | |  | | |__| | | / ____ \|_|            ###
        ###          |_|  |_|\____/|_|/_/    \_\              ###
        ###                                                   ### 
        ######################################################### 
       
 
        
use_float32 set to True: replacing float64 arrays by float32 arrays to speed up computations...

Successfully loaded view='mirna' group='group1' with N=150 samples and D=184 features...
Successfully loaded view='mrna' group='group1' with N=150 samples and D=200 features...
Successfully loaded view='protein' group='group1' with N=150 samples and D=142 features...


Model options:
- Automatic Relevance Determination prior on the factors: False
- Automatic Relevance Determination prior on the weights: True
- Spike-and-slab prior on the factors: False
- Spike-and-slab prior on the weights: False
Likelihoods:
- View 0 (mirna): gaussian
- View 1 (mrna): gaussian
- View 2 (protein): gaussian




######################################
## Training the model with seed 42 ##
######################################


ELBO before training: -605349.12 

Iteration 1: time=0.02, ELBO=-97986.52, deltaELBO=507362.597 (83.81322127%), Factors=10
Iteration 2: time=0.02, Factors=10
Iteration 3: time=0.02, Factors=10
Iteration 4: time=0.02, Factors=10
Iteration 5: time=0.02, Factors=10
Iteration 6: time=0.02, ELBO=-85812.38, deltaELBO=12174.138 (2.01109376%), Factors=10
Iteration 7: time=0.02, Factors=10
Iteration 8: time=0.02, Factors=10
Iteration 9: time=0.02, Factors=10
Iteration 10: time=0.02, Factors=10
Iteration 11: time=0.02, ELBO=-85546.37, deltaELBO=266.018 (0.04394451%), Factors=10
Iteration 12: time=0.01, Factors=10
Iteration 13: time=0.02, Factors=10
Iteration 14: time=0.02, Factors=10
Iteration 15: time=0.02, Factors=10
Iteration 16: time=0.02, ELBO=-85119.84, deltaELBO=426.529 (0.07046008%), Factors=10
Iteration 17: time=0.02, Factors=10
Iteration 18: time=0.02, Factors=10
Iteration 19: time=0.02, Factors=10
Iteration 20: time=0.02, Factors=10
Iteration 21: time=0.02, ELBO=-84663.08, deltaELBO=456.760 (0.07545390%), Factors=10
Iteration 22: time=0.02, Factors=10
Iteration 23: time=0.02, Factors=10
Iteration 24: time=0.02, Factors=10
Iteration 25: time=0.02, Factors=10
Iteration 26: time=0.02, ELBO=-84469.18, deltaELBO=193.893 (0.03202997%), Factors=10
Iteration 27: time=0.02, Factors=10
Iteration 28: time=0.03, Factors=10
Iteration 29: time=0.02, Factors=10
Iteration 30: time=0.02, Factors=10
Iteration 31: time=0.03, ELBO=-84312.91, deltaELBO=156.279 (0.02581635%), Factors=10
Iteration 32: time=0.02, Factors=10
Iteration 33: time=0.02, Factors=10
Iteration 34: time=0.02, Factors=10
Iteration 35: time=0.02, Factors=10
Iteration 36: time=0.02, ELBO=-84211.61, deltaELBO=101.293 (0.01673301%), Factors=10
Iteration 37: time=0.01, Factors=10
Iteration 38: time=0.01, Factors=10
Iteration 39: time=0.01, Factors=10
Iteration 40: time=0.01, Factors=10
Iteration 41: time=0.01, ELBO=-84149.56, deltaELBO=62.055 (0.01025119%), Factors=10
Iteration 42: time=0.01, Factors=10
Iteration 43: time=0.01, Factors=10
Iteration 44: time=0.01, Factors=10
Iteration 45: time=0.01, Factors=10
Iteration 46: time=0.01, ELBO=-84113.44, deltaELBO=36.119 (0.00596669%), Factors=10
Iteration 47: time=0.01, Factors=10
Iteration 48: time=0.01, Factors=10
Iteration 49: time=0.01, Factors=10
Iteration 50: time=0.02, Factors=10
Iteration 51: time=0.02, ELBO=-84094.13, deltaELBO=19.310 (0.00318991%), Factors=10
Iteration 52: time=0.02, Factors=10
Iteration 53: time=0.02, Factors=10
Iteration 54: time=0.02, Factors=10
Iteration 55: time=0.02, Factors=10
Iteration 56: time=0.02, ELBO=-84083.06, deltaELBO=11.071 (0.00182882%), Factors=10
Iteration 57: time=0.02, Factors=10
Iteration 58: time=0.02, Factors=10
Iteration 59: time=0.02, Factors=10
Iteration 60: time=0.02, Factors=10
Iteration 61: time=0.02, ELBO=-84075.86, deltaELBO=7.195 (0.00118864%), Factors=10
Iteration 62: time=0.02, Factors=10
Iteration 63: time=0.02, Factors=10
Iteration 64: time=0.02, Factors=10
Iteration 65: time=0.02, Factors=10
Iteration 66: time=0.02, ELBO=-84070.67, deltaELBO=5.193 (0.00085787%), Factors=10
Iteration 67: time=0.02, Factors=10
Iteration 68: time=0.02, Factors=10
Iteration 69: time=0.02, Factors=10
Iteration 70: time=0.02, Factors=10
Iteration 71: time=0.02, ELBO=-84066.69, deltaELBO=3.977 (0.00065695%), Factors=10
Iteration 72: time=0.02, Factors=10
Iteration 73: time=0.01, Factors=10
Iteration 74: time=0.02, Factors=10
Iteration 75: time=0.02, Factors=10
Iteration 76: time=0.02, ELBO=-84063.53, deltaELBO=3.158 (0.00052174%), Factors=10
Iteration 77: time=0.02, Factors=10
Iteration 78: time=0.01, Factors=10
Iteration 79: time=0.01, Factors=10
Iteration 80: time=0.01, Factors=10
Iteration 81: time=0.01, ELBO=-84060.96, deltaELBO=2.577 (0.00042564%), Factors=10
Iteration 82: time=0.01, Factors=10
Iteration 83: time=0.01, Factors=10
Iteration 84: time=0.01, Factors=10
Iteration 85: time=0.01, Factors=10
Iteration 86: time=0.02, ELBO=-84058.84, deltaELBO=2.121 (0.00035040%), Factors=10

Converged!



#######################
## Training finished ##
#######################


Saving model in /var/folders/hw/jx67_4vj6ljfd13xsg7xzvt83k7mrx/T//Rtmplep0xV/mofa_20231105-162058.hdf5...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .quality_control(object, verbose = verbose): Factor(s) 4 are strongly correlated with the total number of expressed features for at least one of your omics. Such factors appear when there are differences in the total 'levels' between your samples, *sometimes* because of poor normalisation in the preprocessing steps.</code></pre>
</div>
</div>
<p>Note that we get a message that a model was automatically save to .hdf5. It is also possible to specify the name and location of the file to save the model to, via <code>outfile</code> option under <code>runMofa()</code> function.</p>
</section>
<section id="variance-decomposition" class="level2">
<h2 class="anchored" data-anchor-id="variance-decomposition">Variance decomposition</h2>
<p>The most important insight that MOFA generates is the <strong>variance decomposition analysis</strong>. This plot shows the percentage of variance explained by each factor across each data modality.</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-13_fdcd125935b20ce14baac0657b1797a4">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_variance_explained</span>(MOFAobject)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>From the results of the <code>plot_variance_explained</code> function in MOFA, we can discern the variance explained by each factor across the three views: mirna, mrna, and protein.</p>
<p>In the <strong>mirna</strong> view, Factor1 leads by explaining approximately 15.96% of the variance. Notably, Factor1 also stands out in both the <strong>mrna</strong> and <strong>protein</strong> views, explaining 20.37% and 20.41% respectively, suggesting its consistent importance across all views.</p>
<p>For the <strong>mrna</strong> view, besides Factor1, Factor2 contributes significantly with 11.88%. This contrasts with its contribution in the protein view, where it explains only 1.25% of the variance, and in the mirna view, where it accounts for 6.04%.</p>
<p>In the <strong>protein</strong> view, while Factor1 remains dominant, Factor3 emerges as significant, explaining 12.20% of the variance. This is intriguing as Factor3 has a minimal role in the mrna view (0.12%) but does have a presence in the mirna view with 0.65%.</p>
<p>Factors such as Factor4 and Factor7 exhibit diverse roles across the views. In the mirna view, Factor4 explains a notable 12.77% but diminishes to 0.16% and 0.02% in the mrna and protein views respectively. Factor7, on the other hand, is more prominent in the mirna view with 7.09% but is almost negligible in the other two views.</p>
<p>Btw. the variance explained estimates, corresponding to the above plot, are stored in the hdf5 file and loaded in model@cache[[“variance_explained”]]. They can be viewed via:</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-14_32cc827f421ca4b787146e0cf9d135a7">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># variance explained for every factor </span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(MOFAobject<span class="sc">@</span>cache<span class="sc">$</span>variance_explained<span class="sc">$</span>r2_per_factor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$group1
               mirna        mrna     protein
Factor1  15.95515609 20.36821842 20.41299939
Factor2   6.04512691 11.88386679  1.25099421
Factor3   0.64774156  0.12200475 12.20212579
Factor4  12.77297139  0.16268492  0.02241731
Factor5   3.50421071  5.50275445  1.06747746
Factor6   3.08054090  3.41790318  2.97962427
Factor7   7.09193349  0.05609989  0.02453923
Factor8   1.25472546  4.96560931  0.72206855
Factor9   0.01165271  0.11986494  6.47216439
Factor10  2.66354084  2.11876631  0.31054020</code></pre>
</div>
</div>
<p><strong>Which factor consistently plays a vital role across all the views?</strong></p>
<p>It is also important to see if the model fits the data well. We can do this by looking at how much of total variance is explained by factors across different views. Here, the results will usually vary based on the kind of data, number of samples and number of factors used.</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-15_79ef202ecf2de2a70fd015e6e91e0763">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>var_plot <span class="ot">&lt;-</span> <span class="fu">plot_variance_explained</span>(MOFAobject, <span class="at">plot_total =</span> T)[[<span class="dv">2</span>]]</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>var_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>In this data set, by using 10 factors the MOFA model explains about 49% of the variation in the miRNA, 48% of the variation in the mRNA data and 45% of the variation in the protein data.</p>
<p>The matching variance values can be extracted via:</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-16_a7a6672cb3f01bb2a789baa57f6a56bb">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(MOFAobject<span class="sc">@</span>cache<span class="sc">$</span>variance_explained<span class="sc">$</span>r2_total)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$group1
   mirna     mrna  protein 
49.21639 48.26142 44.95106 </code></pre>
</div>
</div>
</section>
<section id="downstream-analysis" class="level2">
<h2 class="anchored" data-anchor-id="downstream-analysis">Downstream analysis</h2>
<p>We now we have a reasonable model in which Factor1 consistently plays a vital role across all views. In the downstream analysis we can explore more factors, such as Factor1, and features weights. We can for instance aim to characterize Factor1 molecular signal and its association with available sample covariates.</p>
</section>
<section id="adding-annotations" class="level2">
<h2 class="anchored" data-anchor-id="adding-annotations">Adding annotations</h2>
<p>We can add samples annotations to the model now. We will add only our cancer subtypes, but could also include other covariates that my be relevant to the study, such as age or gender.</p>
<div class="cell" data-hash="mofa-light_cache/html/unnamed-chunk-17_a8a4eefaedf1ab58bb689fc25ddeb040">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add cancer subtype to the model</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">samples_metadata</span>(MOFAobject) <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">sample=</span><span class="fu">colnames</span>(data_mofa<span class="sc">$</span>mirna),</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">subtype=</span>breast.TCGA<span class="sc">$</span>data.train<span class="sc">$</span>subtype)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualizing-factors" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-factors">Visualizing factors</h3>
<p>We can visualize individual factor or factors combinations. We can also do that in connection to the sample groups. Let’s look at few examples:</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-18_fdd5e0513c79c38c38a42c8415ac5695">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize Factors 1, 2, 3</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> MOFAobject</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_factor</span>(model, </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">factor =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">color_by =</span> <span class="st">"subtype"</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Visualizaiton of individual Factors 1, 2 and 3, with cancer subtype grouping information.</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-19_da1f2388d3ef80d76c27cb60ebf1c7e4">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add violin plot</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> MOFAobject</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_factor</span>(model, </span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">factor =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">color_by =</span> <span class="st">"subtype"</span>, </span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">add_violin =</span> T, </span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dodge =</span> T</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Visualizaiton of individual Factors 1, 2 and 3, with cancer subtype grouping information and with added violin plot.</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-20_f6e0421efee47b2545edfe6567b88b54">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize combination of Factors 1 and 2</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> MOFAobject</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_factors</span>(model, </span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">factors =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">color_by =</span> <span class="st">"subtype"</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Visualizaiton of combinations of Factors 1 and 2, with cancer subtype grouping information.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="association-analysis" class="level3">
<h3 class="anchored" data-anchor-id="association-analysis">Association analysis</h3>
<p>To understand the relation between Factors and sample metadata, we can further perform an association analysis.</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-21_ced30a9a9e9f26449a0faa478967836b">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># correlate factors with covariates</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">correlate_factors_with_covariates</span>(MOFAobject, </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">covariates =</span> <span class="fu">c</span>(<span class="st">"subtype"</span>), </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot =</span> <span class="st">"log_pval"</span>,<span class="at">cluster_cols=</span>F</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in correlate_factors_with_covariates(MOFAobject, covariates =
c("subtype"), : There are non-numeric values in the covariates data.frame,
converting to numeric...</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The results clearly shows a strong association of Factor1 and cancer subtype. The remaining factors do not show a clear association with the cancer subtype. We could have also started from the association analysis to find out the factor that is associated with our grouping or other covariate of interest, and the focus on plotting the factors of interest.</p>
</section>
<section id="visualizing-weights" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-weights">Visualizing weights</h3>
<p>We have a strong trend of subtypes captured by Factor 1. We can now look at the weights for this factor to figure out what are the most important features that contribute to this pattern.</p>
<p>Feature weights play an important role in understanding the influence of each feature on a given factor. These weights offer a quantifiable score for every feature in relation to its respective factor. Essentially, when a feature doesn’t correlate with a factor, its weight is anticipated to hover around zero. Conversely, features that are robustly associated with the factor will display large absolute weight values. The polarity of the weight whether positive or negative reveals the nature of the relationship: a positive weight suggests that the feature level elevates in instances with positive factor values and the opposite for negative weights.</p>
<p>Let’s look at the top 10 features in mRNA.</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-22_fffcf4e0cd7a37b75cfa66680eb09c68">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_top_weights</span>(MOFAobject,<span class="at">view =</span> <span class="st">"mrna"</span>,</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a> <span class="at">factor =</span> <span class="dv">1</span>,</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a> <span class="at">nfeatures =</span> <span class="dv">10</span>,    </span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a> <span class="at">scale =</span> T          </span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The plot suggest that <code>STC2</code> has a strong <em>negative</em> relationship with Factor1. Looking back at the score plot, we see that our <code>Basal</code> subtype has ended up on the right of the plot, <code>Her2</code> in the middle and <code>LumA</code> on the left. This suggest that the expression of <code>STC2</code> is higher in <code>LumA</code> vs <code>Her2</code> and also <code>Her2</code> vs <code>LumA</code>. Let’s check it:</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-23_f9484936706255925362a38211f32aa0">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_data_scatter</span>(MOFAobject, </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">view =</span> <span class="st">"mrna"</span>,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">factor =</span> <span class="dv">1</span>, <span class="at">features =</span> <span class="st">"STC2"</span>,<span class="at">color_by =</span> <span class="st">"subtype"</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Great. But we have so many other features, do we have a subgroup of features in our data:</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-24_ce677adeaefffb008e1bcf251f7b0c71">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot heatmap</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>sample_group <span class="ot">&lt;-</span> <span class="fu">samples_metadata</span>(MOFAobject)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sample_group) <span class="ot">&lt;-</span> sample_group[,<span class="dv">1</span>]</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>heatmap_plot <span class="ot">&lt;-</span> <span class="fu">plot_data_heatmap</span>(MOFAobject, </span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="at">view =</span> <span class="st">"mrna"</span>,</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">factor =</span> <span class="dv">1</span>, <span class="at">features =</span> <span class="dv">50</span>,</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">cluster_rows =</span> <span class="cn">TRUE</span>, <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,<span class="at">annotation_samples =</span> sample_group[,<span class="st">"subtype"</span>,<span class="at">drop=</span>F],</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">show_rownames =</span> <span class="cn">TRUE</span>, <span class="at">show_colnames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">scale =</span> <span class="st">"row"</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'annotation_samples' provided as a data.frame, please make sure that the rownames match the sample names</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>heatmap_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can at least see two big groups of mRNAs having contrasting expression pattern. Let’s extract these features and investigate them further.</p>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-25_cc9d68373c5db7a2e43db6e49696d79a">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cut into features in two groups</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>feature_subgroups <span class="ot">&lt;-</span> <span class="fu">cutree</span>(heatmap_plot<span class="sc">$</span>tree_row, <span class="dv">2</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="co"># plot first group of extracted features</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_data_scatter</span>(MOFAobject, </span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">view =</span> <span class="st">"mrna"</span>,</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">factor =</span> <span class="dv">1</span>,  </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="fu">names</span>(feature_subgroups[feature_subgroups<span class="sc">==</span><span class="dv">1</span>]),</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">color_by =</span> <span class="st">"subtype"</span>,</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">dot_size =</span> <span class="fl">1.5</span></span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="1344"></p>
<figcaption class="figure-caption">Scatter plot of features (mRNAs) in the first group of expression pattern agaist Factor 1 values</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="mofa-light_cache/html/unnamed-chunk-26_5cff42fc49a068a9c8806fe74f5bb556">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot second group of extracted features</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_data_scatter</span>(MOFAobject, </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">view =</span> <span class="st">"mrna"</span>,</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">factor =</span> <span class="dv">1</span>,  </span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> <span class="fu">names</span>(feature_subgroups[feature_subgroups<span class="sc">==</span><span class="dv">2</span>]),</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">color_by =</span> <span class="st">"subtype"</span>, </span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dot_size =</span> <span class="fl">1.5</span></span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="mofa-light_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="1344"></p>
<figcaption class="figure-caption">Scatter plot of features (mRNAs) in the second group of expression pattern agaist Factor 1 values</figcaption>
</figure>
</div>
</div>
</div>
<p>As it is clear the two subgroups are related to the features that are positively and negatively correlated with the first factor. This is a good indication that we can use the weights to group the features. We could use these groups to do enrichment analysis or similar, but this would be outside MOFA package.</p>
<div id="exr-mofa" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 2 (MOFA) </strong></span>Can you perform MOFA on the test data which is in <code>breast.TCGA$data.test</code>?</p>
<ol type="1">
<li>Do you see the same pattern as in the training set?</li>
<li>Do the top 10 most important features overlap between training and testing?</li>
<li>How about the grouping of the features?</li>
</ol>
</div>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>